src/transformStateWithClones.js

```diff
@@ -7,7 +7,34 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateAtTheMoment = { ...state }; // clone the state at the beginning
+  const stateHistory = []; // const, because an array is an object in JS
+
+  for (const action of actions) {
+    let newState = { ...stateAtTheMoment };
+    // Clone the current state to the new state
+
+    if (action.type === 'clear') {
+      newState = {};
+    }
+
+    if (action.type === 'addProperties') {
+      Object.assign(newState, action.extraData);
+    }
+
+    if (action.type === 'removeProperties') {
+      for (const key of action.keysToRemove) {
+        delete newState[key];
+      }
+    }
```

Comment:
Instead of using multiple `if` statements, a `switch` statement would be more appropriate here since you have a limited amount of conditions that are based on the `action.type`. This will also help in ensuring that you have a `default` case for error handling.